File: ruby/difference-of-squares/difference_of_squares.rb
class Squares
  def initialize(number)
    @number = number
  end
  def square_of_sum
    (1..@number).sum ** 2
  end
  def sum_of_squares
    (1..@number).map { |n| n ** 2 }.sum
  end
  def difference
    square_of_sum - sum_of_squares
  end
end

module BookKeeping
  VERSION = 4
end

File: go/difference-of-squares/difference_of_squares.go
package diffsquares

// SumOfSquares (1*1, 2*2, 3*3, ...)
func SumOfSquares(n int) int {
	sum := 0
	for i := range createArrayToNumber(n) {
		sum += i * i
	}
	return sum
}

// SquareOfSums (1 + 2 + ... + 10)2
func SquareOfSums(n int) int {
	sum := 0
	for i := range createArrayToNumber(n) {
		sum += i
	}
	return sum * sum
}

// Difference between the square of the sum and the sum of the squares of the first (n) natural numbers
func Difference(n int) int {
	return SquareOfSums(n) - SumOfSquares(n)
}

func createArrayToNumber(n int) []int {
	ary := make([]int, n+1)
	for i := 1; i < n+1; i++ {
		ary[i] = i
	}
	return ary
}
